File: jiho/level1/오픈채팅방.java
//https://school.programmers.co.kr/learn/courses/30/lessons/42888

package level1;

import java.util.*;

class 오픈채팅방 {
    public String[] solution(String[] record) {
        // 내용(뒷부분) 저장
        List<String> chat = new ArrayList<>();
        // 최신 닉네임(앞부분) 저장
        HashMap<String, String> nickNameMap = new HashMap<>();

        for(int i = 0; i<record.length; i++){
            StringTokenizer st = new StringTokenizer(record[i]);
            String cmd = st.nextToken();
            String id = st.nextToken();
            String nickName = "";

            if(cmd.equals("Enter") || cmd.equals("Change")){
                nickName = st.nextToken();
            }

            switch (cmd){
                case "Enter" :
                    nickNameMap.put(id, nickName);
                    chat.add(id + "님이 들어왔습니다.");
                    break;
                case "Leave" :
                    chat.add(id + "님이 나갔습니다.");
                    break;
                case "Change" :
                    nickNameMap.put(id, nickName);
                    break;
            }
        }

        String[] answer = new String[chat.size()];
        int i = 0;

        for(String str : chat){
            int index = str.indexOf("님");
            String id = str.substring(0,index);

            answer[i++] = str.replace(id, nickNameMap.get(id));
        }

        return answer;
    }
}

File: ningpop/level2/오픈채팅방.py
# 2022.08.04
# 풀이 시간: 21분 42초
# 채점 결과: 정답
# 시간복잡도: O(N)
# 문제 링크: https://school.programmers.co.kr/learn/courses/30/lessons/42888

def solution(record: list) -> list:
    user_table = {}
    result = []
    for data in record:
        splited_data = list(data.split())
        order = splited_data[0]
        uid = splited_data[1]

        if order == 'Enter':
            nickname = splited_data[2]
            user_table[uid] = nickname
            result.append(('Enter', uid))
        elif order == 'Leave':
            nickname = user_table[uid]
            result.append(('Leave', uid))
        elif order == 'Change':
            nickname = splited_data[2]
            user_table[uid] = nickname

    answer = []
    for data in result:
        order, uid = data[0], data[1]

        if order == 'Enter':
            answer.append(f'{user_table[uid]}님이 들어왔습니다.')
        elif order == 'Leave':
            answer.append(f'{user_table[uid]}님이 나갔습니다.')
    
    return answer

print(solution(
    ["Enter uid1234 Muzi", "Enter uid4567 Prodo","Leave uid1234","Enter uid1234 Prodo","Change uid4567 Ryan"]
)) # ["Prodo님이 들어왔습니다.", "Ryan님이 들어왔습니다.", "Prodo님이 나갔습니다.", "Prodo님이 들어왔습니다."]